### DISCUSSION TITLE
How to view question.html in the backend?

### DISCUSSION THREAD

**Discussion by aalrabah** [PrairieLearn/PrairieLearn#10072] [Dev]
I'm working on a PrairieLearn project(running docker locally) and encountering an issue with accessing question.html content.

Environment:
  Backend: grade/run directory.
  Docker: Can view /course/questions and question.html exists.
  Frontend: Displays question correctly, retrieves solution and user input in the backend. 
  Server: Up and running; everything is working fine.

How can I retrieve the content of question.html from the backend given the current directory structure and Docker setup?

**Comment by jonatanschroeder** [comment]
Can you clarify what you mean by "backend"? Do you mean server.py, or an external grader, or a workspace, or something else?

**Reply by aalrabah** [reply]
By "backend," I mean the code responsible for grading the questions, specifically the grader in the serverFilesCourse. This code is in my local environment and is mounted through Docker. I'm trying to access the question.html file from the /course/questions directory within this backend setup

**Reply by nwalters512** [reply]
https://xyproblem.info/ - can you say more about your situation and what you're actually trying to achieve?

**Reply by aalrabah** [reply]
I'm trying to retrieve the content from a given question page on the front end. In my grader function, I already save the solution output and user input (answer) into a JSON file. On that same page, I want to also save the problem statement, which is found in question.html. However, I'm not sure how question.html is being rendered or handled in PrairieLearn, and thus I'm unable to retrieve its content. Any guidance on accessing question.html would be greatly appreciated.

**Reply by nwalters512** [reply]
I'm going to dig a little deeper here before recommending any solution to make sure you're doing the best thing for your scenario. How do you plan on using the question HTML in your external grader?

**Reply by jonatanschroeder** [reply]
I'm going to go on a limb and guess that what you're trying to do is to retrieve the rendered question panel, not the raw question.html itself, is that correct? More precisely, you want the dynamic portion of the question panel, which you generated in the generate function of server.py, is my assumption correct?

If that is indeed what you want, note that you can retrieve params, correct_answers and other relevant values [from `/grade/data/data.json`](https://prairielearn.readthedocs.io/en/latest/externalGrading/#the-grading-process).

 If this is not the case, [please describe why you need the question file and what information you're hoping to retrieve from it](https://mywiki.wooledge.org/XyProblem).

**Reply by aalrabah** [reply]
@nwalters512  @jonatanschroeder 

We need to retrieve three pieces of information: the user's input, the text of the question for which the student submitted a response, and the correct answer for the current question. Then, we need to make an external API call that takes these three parameters as input (in JSON format). The API call will return a string value, which we want to display to the user on the frontend.

Currently, we have access to the user's input and the correct answer. However, we are unable to determine which question the answer is submitted for, and we cannot access the questions list within the code. Additionally, our serverFilesCourse is restricted to an isolated copied directory, limiting our ability to browse and access necessary files

**Reply by jonatanschroeder** [reply]
Any information from a question that you want to have available in the external grader should be provided either in the `tests` directory or in the params/correct answer from server.py. Among other things this is done to restrict the amount of files that are available in the testing environment that need to be sandboxed by this environment.

My suggestion is that you provide the question statement in a different location (not directly in question.html), and then retrieve it from that source from both question.html and the external grader. Two options come to mind (others may be possible):

* If the question is relatively short (one or two statements), you can set the question statement in server.py to something like `data["params"]["question_statement"]`. This would allow you to set it in question.html using `{{params.question_statement}}` (or `{{{params.question_statement}}} if you include formatting), and to retrieve it from the external grader using `/grade/data/data.json`.
* If the question is slightly longer, you could save the question statement as a file (text, HTML, JSON or whatever format makes sense for you) inside the question's `tests` directory. This would make it very easy to retrieve from the external grader from `/grade/tests`. As for question.html, you can create a custom element in your course that reads that file and renders its content based on the retrieved information.

I hope this helps.

**Reply by aalrabah** [reply]
@jonatanschroeder 

Thank you for that that was helpful, I was able to retrieve the questions through suggestion 2. :)
